<a href="https://colab.research.google.com/github/saurabhd1008/barrister-ai/blob/main/Copy_of_Session_4_Part_4_Embeddings.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Install the Sentence Transformer library
!pip install --quiet --upgrade sentence-transformers

import os
import logging

# Suppress Hugging Face / Transformers logs
os.environ["TRANSFORMERS_VERBOSITY"] = "error"  # only show errors

logging.getLogger("transformers").setLevel(logging.ERROR)
logging.getLogger("sentence_transformers").setLevel(logging.ERROR)

from sentence_transformers import SentenceTransformer, util

# Load model silently
model = SentenceTransformer("sentence-transformers/all-MiniLM-L6-v2")

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 588.9/588.9 kB 7.6 MB/s eta 0:00:00


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [ ]:
def get_embedding(text):
    return model.encode(text,convert_to_tensor=True)

def find_text_similarity(text1, text2):
    return round(float(util.pytorch_cos_sim(get_embedding(text1), get_embedding(text2))[0][0]), 4)

def find_embedding_similarity(emb1, emb2):
    return round(float(util.pytorch_cos_sim(emb1, emb2)[0][0]), 4)

# LLM Tools

In [ ]:
text = "I am a human being."
text_embedding = get_embedding(text)
print(len(text_embedding))
print(text_embedding)

# An embedding is a vector of numbers
# which captures the meaning of a given sentence in an abstract way.
# We cannot understand these numbers ourselves.
# But these are very helpful in many ML and AI applications.
# These embeddings basically form the input for all ML applications,
# and are a non-explainable substitute for feature engineering.

384
tensor([ 5.2786e-03,  1.9185e-02,  6.3271e-02,  1.1727e-02, -5.7958e-02,
        -7.7475e-02,  1.4263e-01, -3.2002e-02,  4.6559e-02,  9.9696e-03,
        -9.2031e-03, -9.0321e-02, -3.2316e-02, -1.8097e-02,  7.7099e-02,
        -3.1339e-02, -5.2647e-03, -4.3888e-02, -4.0001e-02,  4.4545e-02,
        -2.9328e-02,  4.6565e-02, -1.5779e-02, -2.0519e-02, -4.9357e-02,
        -1.5548e-02, -1.0629e-02, -3.4567e-02,  6.5551e-02, -1.7427e-02,
         1.2922e-02, -6.9742e-02,  9.9495e-02, -1.9190e-02, -3.6043e-02,
         7.4612e-02,  4.4963e-02, -5.4321e-02,  5.6025e-02,  1.8090e-02,
         3.3005e-02, -8.2020e-02, -6.1142e-03,  5.9380e-03,  3.7282e-02,
         5.0122e-02, -2.1457e-02, -4.2981e-02, -4.2478e-02, -5.2496e-02,
        -7.1995e-02,  1.0558e-01, -3.2934e-02,  6.6258e-02, -8.5375e-03,
        -4.0289e-02,  4.9874e-02,  2.4564e-02,  2.1757e-02, -5.7993e-02,
         9.0196e-03,  9.0460e-03, -1.8721e-02,  1.0411e-01,  4.4548e-02,
        -8.3939e-03,  3.7495e-02, -4.1318e-02, 

In [ ]:
text_embedding.shape

torch.Size([384])

In [ ]:
text1 = "I am a human being."
text2 = "What are you doing?"
find_text_similarity(text1, text2)
# So this value ranges between -1 to +1.

# ONLY TEXT DOES NOT HAVE A NATURAL NUMERIC REPRESENTATION.

0.3407

# Word Embedding Plots

In [ ]:
# Define word list
words = [
    "apple", "banana", "mango", "orange",
    "tiger", "lion", "cat", "dog",
    "car", "truck", "bicycle", "train",
    "happy", "sad", "angry", "excited"
]

# Define categories (same order as words)
categories = [
    "fruit", "fruit", "fruit", "fruit",
    "animal", "animal", "animal", "animal",
    "vehicle", "vehicle", "vehicle", "vehicle",
    "emotion", "emotion", "emotion", "emotion"
]

In [ ]:
# Get embeddings
embeddings = model.encode(words)

In [ ]:
# Reduce to 2D using PCA
# from sklearn.decomposition import PCA
# pca = PCA(n_components=2)
# points = pca.fit_transform(embeddings)

# # Use TSNE
from sklearn.manifold import TSNE
points = TSNE(n_components=2, perplexity=5, random_state=42).fit_transform(embeddings)

In [ ]:
# Create a DataFrame for Plotly
import pandas as pd
df = pd.DataFrame({
    'word': words,
    'category': categories,
    'x': points[:,0],
    'y': points[:,1]
})

In [ ]:
import plotly.express as px

# Interactive scatter plot
fig = px.scatter(
    df,
    x='x',
    y='y',
    color='category',
    text='word',
    title="🧩 Word Embeddings in 2D — Similar Words Stick Together!",
    color_discrete_sequence=px.colors.qualitative.Pastel
)

# Style and show
fig.update_traces(textposition='top center', marker=dict(size=12, opacity=0.8))
fig.update_layout(
    width=800,
    height=600,
    showlegend=True,
    template='plotly_white'
)

fig.show()

**DIY:**

1. Take the Brown corpus dataset, and find embeddings for each of the 500 texts using the above SBERT model. This will give you a vector of size 384 for each of the texts in this corpus.

2. Apply Logistic Regression using these embeddings as features.

3. Do K-Means clustering of these embeddings, and analyse how many texts in each cluster belong to the same class.